rust/bridge/node/futures/src/future.rs

```diff
@@ -0,0 +1,205 @@
+    Complete(std::thread::Result<T>),
+    /// The result has been returned from `poll`.
+    Consumed,
+}
+
+impl<T> JsFutureState<T> {
+    fn new(transform: impl JsFutureCallback<T>) -> Self {
+        Self::Waiting {
+            transform: Box::new(transform),
+            waker: None,
+        }
+    }
+
+    fn waiting_on(mut self, new_waker: Waker) -> Self {
+        if let Self::Waiting { ref mut waker, .. } = self {
+            *waker = Some(new_waker)
+        } else {
+            panic!("already completed")
+        }
+        self
+    }
+}
+
+/// A future representing the result of a JavaScript promise.
+///
+/// JsFutures can be created with [await_promise](fn@JsFuture::await_promise) or [get_promise](fn@JsFuture::get_promise) and [JsFutureBuilder].
+/// Once resolved, a transformation callback is invoked in the current JavaScript context to produce a Rust value.
+/// This is the result of `await`ing the future.
+///
+/// Panics in the transformation function will be propagated to the `await`ing context.
+pub struct JsFuture<T: 'static + Send> {
+    // In practice there will only be one strong reference to one of these from Rust,
+    // and two weak ones from JavaScript (for `resolve` and `reject`).
+    shared: Arc<Mutex<JsFutureState<T>>>,
+}
+
+impl<T: 'static + Send> Future for JsFuture<T> {
+    type Output = T;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut std::task::Context) -> Poll<Self::Output> {
+        let mut state_guard = self.shared.lock().unwrap();
+        let state = mem::replace(&mut *state_guard, JsFutureState::Consumed);
+        match state {
+            JsFutureState::Complete(Ok(result)) => return Poll::Ready(result),
+            JsFutureState::Complete(Err(panic)) => resume_unwind(panic),
+            JsFutureState::Consumed => panic!("already consumed"),
+            JsFutureState::Waiting { .. } => {}
+        }
+        *state_guard = state.waiting_on(cx.waker().clone());
+        Poll::Pending
+    }
+}
+
+struct WeakFutureToken<T: Send + 'static>(Weak<Mutex<JsFutureState<T>>>);
+
+impl<T: Send + 'static> WeakFutureToken<T> {
+    /// Creates a token referencing `future`.
+    fn new(future: &JsFuture<T>) -> Self {
+        Self(Arc::downgrade(&future.shared))
+    }
+
+    /// Produces a JavaScript function value representing [fulfill_promise]
+    /// with the first argument bound to this token, boxed.
+    fn bind_fulfill_promise<'a, C: Context<'a>, R: JsPromiseResultConstructor>(
+        &self,
+        cx: &mut C,
+    ) -> JsResult<'a, JsValue> {
+        let fulfill = JsFunction::new(cx, fulfill_promise::<T, R>)?;
+        let bind = fulfill
+            .get(cx, "bind")?
+            .downcast_or_throw::<JsFunction, _>(cx)?;
+        let bind_args = vec![
+            cx.undefined().upcast::<JsValue>(),
+            cx.boxed(self.clone()).upcast(),
+        ];
+        bind.call(cx, fulfill, bind_args)
```

Comment:
Unfortunately Weak doesn't implement [Finalize](https://docs.rs/neon/0.7.0-napi.3/neon/prelude/trait.Finalize.html), so we need *some* custom type (or a wrapper that says "no custom finalization needed").

I'm also not sure about pulling out `bound_js_function`, because in this case it's binding the argument and in `promise` it's binding the `this` value. I'll look and see if anywhere else can make use of that.